### ISSUE TITLE
Write code snippets for common navigation scenarios

### ISSUE THREAD

**Issue by InMatrix** [flutter/uxr#9]
To conduct a comparative analysis (#7), we need to write code snippets for common navigation scenarios for each package we'd like to examine. To do that, we will create a folder `nav2-usability` in this repo to host the snippets. The folder is likely to be organized as follows:

- nav2-usability
  - scenario-code
    - deeplink-pathparam
      - deeplink-pathparam-scenario.md
      - deeplink-pathparam-storyboard.png
      - deeplink_pathparam_{package_name}.dart
      - deeplink_pathparam_{another_package_name}.dart
    - {another-scenario} 

Contributions are welcomed from developers who're familiar with any of the packages we examine.  

TODO List:

- [X] Code snippets for Router 
- [x] Code snippets for vrouter – missing nested routing
- [x] Code snippets for beamer
- [x] Code snippets for auto_route

To make progress as [planned](https://github.com/flutter/uxr/issues/31#issue-842660244), We need code snippets for at least these 3 scenarios: deep linking by path parameters, sign-in, and nested routing by April 30th.

**Comment by slovnicki** [comment]
A good idea would be to implement scenarios with pure Navigator 2.0 and smallest possible UI whose "theme" spans throughout examples (for example, "books theme", for familiarity). Those would be the base examples.

Then, base examples will serve as a starting point for accomplishing the same functionality with a package, keepeing the UI as intact as possible. This way we'll have a clearer view into the differences a package provides. If we furthermore start each package implementation by copying base examples first, we will even have a more straightforward insight into diffs.

**Comment by InMatrix** [comment]
@slovnicki This is a good idea. I just talked to @johnpryan and others. We are going to look into the feasibility of tying scenarios together under one app theme and then creating snippets with the vanilla Navigator 2.0 API first as the baseline for comparison with snippets written with various packages.

**Comment by idkq** [comment]
> smallest possible UI

Agree as a starter, however, we should not leave normal/complex and more structured examples away before build/delivery. Something like https://github.com/flutter/samples. Not having it was a mistake on Nav 2.0. Reasons: 

1) only real-life scenarios give full hint on the complexities one will face when adopting the model.
2) many users were not able to find good examples with nav 2.0 for seemly standard navigation patterns such as using BottomNavigator (I understand that we have a separate use case for this now but not having solid use cases integrations can definitely lead to missing scenarios).
3) by definition, User Experience depends on a robust "prototype", not only simple book-details examples.
4) if shipped only with simple/basic/minimalistic examples it is likely that we will find some compatibility or integration issue in the future.
5) many developers just look at the code (like it or not) and not documentation, so good examples are really necessary

I suggest we define some complex use cases, maybe open a GitHub issue so people can post ideas for us to put the framework to test (a so-called stress-test). I know its more work but it will ensure it will be done right.

**Comment by nguyenxndaidev** [comment]
When snippets for isolated scenarios are done with the the vanilla Navigator 2.0 API, it would be nice if we can continue with a complete bookstore app with all features. This way, we will push all packages to it's limitations.

**Comment by tomgilder** [comment]
Hey @nguyenxndaidev - I've made one, have been meaning to post about it.

You can find it here: https://github.com/tomgilder/routemaster/tree/main/example/book_store

It should be reasonable simple to adapt it to use other routers. It also has a [set of tests](https://github.com/tomgilder/routemaster/blob/main/example/book_store/test/navigation_test.dart) which are router-agnostic, and could be good for testing other routers work correctly.

Welcome to Dashazon! 😀 

<img width="300" alt="Screenshot 2021-04-05 at 15 32 29" src="https://user-images.githubusercontent.com/756862/113574116-234a6800-9624-11eb-8546-7dc6829a35f1.png">

**Comment by esDotDev** [comment]
> I suggest we define some complex use cases, maybe open a GitHub issue so people can post ideas for us to put the framework to test (a so-called stress-test). I know its more work but it will ensure it will be done right.

If we want a complex example, that is closer to a real-world sprawling app, then something like twitter would be a great candidate. I mentioned this in another thread, but it's has many nested tab menus, and all sorts of various deeplinks, it's a great real world case study and relatable to almost everyone.

For the book example, I think a simple example like this could work, but only with very strict requirements, ie:
* State should be retained between tabs (demonstrate state retention between routes)
* A permanent scaffold should exist that does not animate off screen _ever_ (demonstrate scaffold nesting)
* A custom fadeIn/fadeOut page transition should be used (demonstrate custom page transitions)

That would probably push the level of complexity high enough that rough spots / pain points would emerge?

**Comment by InMatrix** [comment]
Now #34 has been merged, I hope there is nothing blocking package authors (and their supporters) from submitting equivalent snippets. 

@slovnicki @lulupointu @theweiweiway, we really need your help completing snippets for `vrouter`, `beamer`, and `auto_route`, so we can carry out the usability evaluations as planned. Would it be possible to get this done by April 30th? Thank you.

**Comment by slovnicki** [comment]
@InMatrix I was planning to submit the rest of scenarios this weekend, so that's a yes from beamer.

**Comment by theweiweiway** [comment]
A - OK @InMatrix-sama

**Comment by InMatrix** [comment]
Thanks, @slovnicki, @theweiweiway, and @lulupointu!

**Comment by theweiweiway** [comment]
@InMatrix is it acceptable to use Provider in the snippets?

**Comment by esDotDev** [comment]
Any routing solutions that use Page Routes, will likely need something like Provider to pass down shared state if they want their routes to 'maintain state' so to speak. My solutution naturally worked around this because it uses a single page route that wrapped an `IndexedStack`, but for the vast majority of routing solutions use full screen routes, so they will need to hoist that persistent state up, and pass it back down. Allowing Provider makes a lot of sense to me, otherwise you'll just get a mix of property drilling or inherited widgets, all adding extra code that is not important or informative.

**Comment by nguyenxndaidev** [comment]
I am not sure why we need Provider, could you please give an example code where we cannot live without Provider?
In my case, I never need Provider for scenarios in this research yet.

**Comment by lulupointu** [comment]
Agreed with @nguyenxndaidev

**Comment by theweiweiway** [comment]
Hey @InMatrix, yeah - I understand why you want to keep everything lean and only the core Flutter API. The only reason I thought Provider might be OK is because it's widely used as a dependency and is the go-to-solution instead of InheritedWidget to pass things down the tree. Anyways, the reason I decided to ask is for the **sign-in scenario** - specifically for signing the user out.

In AutoRoute I can't actually provide the `appState` down to each screen in a direct manner to allow the user to sign out. I know in VRouter you can do:

```dart
class _BooksAppState extends State<BooksApp> {
  final AppState _appState = AppState(MockAuthentication());

  @override
  Widget build(BuildContext context) {
    return VRouter(
      routes: [
        ...
                builder: (context) => HomeScreen(
                  onSignOut: () async {
                    await _appState.signOut(); // access _appState inside of route map
                    context.vRouter.push('/signIn');
                  },
                ),
        ...   
```

But is this realistic? To have your state logic/methods mixed in with the route map? (Sorry to VRouter as an example here lulu). Imagine if you had tens of state methods that you were passing to different screens.

Shouldn't we be trying to mimic real-world scenarios and use-cases with these snippets? The current `router.dart` example defines `handleSignIn` and `handleSignOut` in a central location, which is then passed as arguments to each screen that requires it. This doesn't feel very realistic to me. As a developer, if i need to change the state, I don't think I would define my state mutating methods in a central location, only to be passed to my individual screens via callbacks. Instead, wouldn't the obvious thing to do be to use a provider and then call the state mutating function exactly where I need it? It's simpler, cleaner, easier to understand.

I guess this brings back the whole imperative vs declarative debate and the complicated relationship that navigation and state have..

But overall - what I'm trying to say here is - Is it truly a good idea to be judging and evaluating these packages based on non-realistic scenarios (ie. no access to Provider)? Will this allow you guys to come to an accurate and informed result?

**Comment by theweiweiway** [comment]
And just to clarify - not trying to put down VRouter or lulu at all - I only added it for a comparison. I think VRouter is a great package, and has the flexibility to be written without Provider, as well as **with** Provider too!

**Comment by nguyenxndaidev** [comment]
`final AppState _appState = AppState(MockAuthentication());`

In real life, I will use dependency injection to get the instance of authentication service here.
You can simulate DI by creating a global instance and reuse the instance directly in your widget, so you don't need another package like Provider in this case.

**Comment by johnpryan** [comment]
> Is it truly a good idea to be judging and evaluating these packages based on non-realistic scenarios (ie. no access to Provider)?

InheritedWidget is a realistic state-management solution too, but I think for the purposes of this study, the samples are intended to be small and focused. 

> Will this allow you guys to come to an accurate and informed result?

I think introducing other APIs will distract from the code being studied. The snippets we currently have don't use provider or InheritedWidget, so I hope that other package snippets follow the same pattern.

**Comment by johnpryan** [comment]
I agree with @nguyenxndaidev about global variables - it's a little contrived, but easier to read and understand quickly for participants.

**Comment by theweiweiway** [comment]
Hey @johnpryan

So just to be clear - Provider or InheritedWidget is a no go but  @nguyenxndaidev 's recommendation is good to go, right?

**Comment by johnpryan** [comment]
Yes, please use global variables if possible. If you need to use InheritedWidget make sure you add it to the end of the snippet so it doesn't distract from the code being studied.

**Comment by theweiweiway** [comment]
I see. OK - thanks!

**Comment by InMatrix** [comment]
Closed by #63.
